graph/src/components/ethereum/adapter.rs

```diff
@@ -661,31 +710,129 @@ pub trait EthereumAdapter: Send + Sync + 'static {
     /// reorgs.
     /// It is recommended that `to` be far behind the block number of latest block the Ethereum
     /// node is aware of.
-    fn blocks_with_logs(
-        &self,
-        logger: &Logger,
+    fn blocks_with_triggers(
+        self: Arc<Self>,
+        logger: Logger,
+        chain_store: Arc<dyn ChainStore>,
         subgraph_metrics: Arc<SubgraphEthRpcMetrics>,
         from: u64,
         to: u64,
         log_filter: EthereumLogFilter,
-    ) -> Box<dyn Future<Item = Vec<EthereumBlockPointer>, Error = Error> + Send>;
+        call_filter: EthereumCallFilter,
+        block_filter: EthereumBlockFilter,
+    ) -> Box<dyn Future<Item = Vec<EthereumBlockWithTriggers>, Error = Error> + Send> {
+        // Each trigger filter needs to be queried for the same block range
+        // and the blocks yielded need to be deduped. If any error occurs
+        // while searching for a trigger type, the entire operation fails.
+        let eth = self.clone();
+        let mut trigger_futs: futures::stream::FuturesUnordered<
+            Box<dyn Future<Item = Vec<EthereumTrigger>, Error = Error> + Send>,
+        > = futures::stream::FuturesUnordered::new();
+
+        if block_filter.trigger_every_block {
+            trigger_futs.push(Box::new(self.block_range_to_ptrs(&logger, from, to).map(
+                move |ptrs| {
+                    ptrs.into_iter()
+                        .map(|ptr| EthereumTrigger::Block(ptr, EthereumBlockTriggerType::Every))
+                        .collect()
+                },
+            )))
+        }
 
-    fn blocks_with_calls(
+        // Scan the block range from triggers to find relevant blocks
+        if !log_filter.is_empty() {
+            trigger_futs.push(Box::new(
+                eth.blocks_with_logs(&logger, subgraph_metrics.clone(), from, to, log_filter)
+                    .map(|logs: Vec<Log>| logs.into_iter().map(EthereumTrigger::Log).collect()),
+            ))
+        }
+
+        if !call_filter.is_empty() {
+            trigger_futs.push(Box::new(
+                eth.blocks_with_calls(&logger, subgraph_metrics.clone(), from, to, call_filter)
+                    .map(EthereumTrigger::Call)
+                    .collect(),
+            ));
+        }
+
+        match block_filter.contract_addresses.len() {
```

Comment:
Should we put this one right next to the `block_filter.trigger_every_block` condition above, perhaps even in an `else` branch to make it more obvious that the block filter is only applied once, depending on which type of block filter it is.

Reply:
Good idea, I've refactored this a bit.